aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -562,19 +564,45 @@ impl AptosVM {
     fn validate_publish_request(
         &self,
         modules: &[CompiledModule],
-        expected_names: BTreeSet<String>,
+        mut expected_modules: BTreeSet<String>,
+        allowed_deps: Option<BTreeMap<AccountAddress, BTreeSet<String>>>,
     ) -> VMResult<()> {
-        let given_names = modules
-            .iter()
-            .map(|m| m.self_id().name().as_str().to_string())
-            .collect::<BTreeSet<_>>();
-        if given_names != expected_names {
-            Err(PartialVMError::new(StatusCode::VERIFICATION_ERROR)
-                .with_message("metadata and code bundle mismatch".to_owned())
-                .finish(Location::Undefined))
-        } else {
-            Ok(())
+        for m in modules {
+            if !expected_modules.remove(m.self_id().name().as_str()) {
+                return Err(Self::metadata_validation_error(&format!(
```

Comment:
Do these errors get properly propagated in the transaction failure?

Reply:
Yes, there are tests for this. Actually this PR fixes a bug because previously, a Discard error was produced. Now its a Keep error.